firebase-firestore/src/test/java/com/google/firebase/firestore/local/SQLiteQueryEngineTest.java

```diff
@@ -277,4 +277,94 @@ public void orQueryWithArrayMembershipUsingIndexes() throws Exception {
         expectOptimizedCollectionScan(() -> runQuery(query2, SnapshotVersion.NONE));
     assertEquals(docSet(query2.comparator(), doc1, doc4, doc6), result2);
   }
+
+  @Test
+  public void queryWithMultipleInsOnTheSameField() throws Exception {
+    MutableDocument doc1 = doc("coll/1", 1, map("a", 1, "b", 0));
+    MutableDocument doc2 = doc("coll/2", 1, map("b", 1));
+    MutableDocument doc3 = doc("coll/3", 1, map("a", 3, "b", 2));
+    MutableDocument doc4 = doc("coll/4", 1, map("a", 1, "b", 3));
+    MutableDocument doc5 = doc("coll/5", 1, map("a", 1));
+    MutableDocument doc6 = doc("coll/6", 1, map("a", 2));
+    addDocument(doc1, doc2, doc3, doc4, doc5, doc6);
+    indexManager.addFieldIndex(fieldIndex("coll", "a", Kind.ASCENDING));
+    indexManager.addFieldIndex(fieldIndex("coll", "a", Kind.DESCENDING));
+    indexManager.addFieldIndex(fieldIndex("coll", "b", Kind.ASCENDING));
+    indexManager.addFieldIndex(fieldIndex("coll", "b", Kind.DESCENDING));
+    indexManager.updateIndexEntries(docMap(doc1, doc2, doc3, doc4, doc5, doc6));
+    indexManager.updateCollectionGroup("coll", IndexOffset.fromDocument(doc6));
+
+    // a IN [1,2,3] && a IN [0,1,4] should result in "a==1".
+    Query query1 =
+        query("coll")
+            .filter(
+                andFilters(
+                    filter("a", "in", Arrays.asList(1, 2, 3)),
+                    filter("a", "in", Arrays.asList(0, 1, 4))));
+    DocumentSet result1 =
+        expectOptimizedCollectionScan(() -> runQuery(query1, SnapshotVersion.NONE));
+    assertEquals(docSet(query1.comparator(), doc1, doc4, doc5), result1);
+
+    // a IN [2,3] && a IN [0,1,4] is never true and so the result should be an empty set.
+    Query query2 =
+        query("coll")
+            .filter(
+                andFilters(
+                    filter("a", "in", Arrays.asList(2, 3)),
+                    filter("a", "in", Arrays.asList(0, 1, 4))));
+
+    DocumentSet result2 =
+        expectOptimizedCollectionScan(() -> runQuery(query2, SnapshotVersion.NONE));
+    assertEquals(docSet(query2.comparator()), result2);
+
+    // a IN [0,3] || a IN [0,2] should union them (similar to: a IN [0,2,3]).
+    Query query3 =
+        query("coll")
+            .filter(
+                orFilters(
+                    filter("a", "in", Arrays.asList(0, 3)),
+                    filter("a", "in", Arrays.asList(0, 2))));
```

Comment:
Can we add another case with a more complex composite filter, like the ones in `LogicUtilTest`? Same below.

Reply:
Done